crates/uv-resolver/src/resolver/environment.rs

```diff
@@ -0,0 +1,420 @@
+            .collect()
+    }
+
+    /// Narrow the [`PythonRequirement`] if this resolver environment
+    /// corresponds to a more constraining fork.
+    ///
+    /// For example, if this is a fork where `python_version >= '3.12'` is
+    /// always true, and if the given python requirement (perhaps derived from
+    /// `Requires-Python`) is `>=3.10`, then this will "narrow" the requirement
+    /// to `>=3.12`, corresponding to the marker expression describing this
+    /// fork.
+    ///
+    /// If this environment is not a fork, then this returns `None`.
+    pub(crate) fn narrow_python_requirement(
+        &self,
+        python_requirement: &PythonRequirement,
+    ) -> Option<PythonRequirement> {
+        Some(python_requirement.narrow(&self.requires_python_range()?)?)
+    }
+
+    /// Returns a message formatted for end users representing a fork in the
+    /// resolver.
+    ///
+    /// If this resolver environment does not correspond to a particular fork,
+    /// then `None` is returned.
+    ///
+    /// This is useful in contexts where one wants to display a message
+    /// relating to a particular fork, but either no message or an entirely
+    /// different message when this isn't a fork.
+    pub(crate) fn end_user_fork_display(&self) -> Option<impl std::fmt::Display + '_> {
+        match self.kind {
+            Kind::Specific { .. } => None,
+            Kind::Universal { ref markers, .. } => {
+                if markers.is_true() {
+                    None
+                } else {
+                    Some(format!("split ({markers:?})"))
+                }
+            }
+        }
+    }
+
+    /// Returns the marker expression corresponding to the fork that is
+    /// represented by this resolver environment.
+    ///
+    /// This returns `None` when this does not correspond to a fork.
+    pub(crate) fn try_markers(&self) -> Option<&MarkerTree> {
+        match self.kind {
+            Kind::Specific { .. } => None,
+            Kind::Universal { ref markers, .. } => {
+                if markers.is_true() {
+                    None
+                } else {
+                    Some(markers)
+                }
+            }
+        }
+    }
+
+    /// Returns a requires-python version range derived from the marker
+    /// expression describing this resolver environment.
+    ///
+    /// When this isn't a fork, then there is nothing to constrain and thus
+    /// `None` is returned.
+    fn requires_python_range(&self) -> Option<RequiresPythonRange> {
+        crate::marker::requires_python(self.try_markers()?)
+    }
+}
+
+/// A user visible representation of a resolver environment.
+///
+/// This is most useful in error and log messages.
+impl std::fmt::Display for ResolverEnvironment {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self.kind {
+            Kind::Specific { .. } => write!(f, "Marker Environment"),
```

Comment:
These messages should start with a lower case letter, so that we don't get an upper case letter mid-sentence.